### PR TITLE
refactor(stream): split SortBuffer::iter_until to set_watermark && consume

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

When we set_watermark, we may don't want to consume the output immediately.

And do some little cleaning.

* Remove gen-iter

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/stream/src/executor/sort_buffer.rs [hunk 1]
```diff
@@ -58,12 +57,29 @@ pub struct SortBuffer<S: StateStore> {
...
 }
 
 impl<S: StateStore> SortBuffer<S> {
+    /// Update the watermark value.
+    pub fn set_watermark(&mut self, watermark: ScalarImpl) {
+        assert!(
+            {
+                if let Some(ref last_watermark) = self.watermark {
+                    &watermark >= last_watermark
+                } else {
```
Comment (st1page) [inline src/stream/src/executor/sort_buffer.rs:74]: will we have functions such as `cmp_watermark` when we remove the `Ord` implementation on ScalarImpl?

File: src/stream/src/executor/sort_buffer.rs [hunk 2]
```diff
@@ -129,46 +146,52 @@ impl<S: StateStore> SortBuffer<S> {
...
-            for (_, (row, _)) in self.buffer.range((start_bound, end_bound)) {
-                if let Some(data_chunk) = data_chunk_builder.append_one_row(row) {
-                    // When the chunk size reaches its maximum, we construct a data chunk and
-                    // send it to downstream.
-                    yield data_chunk;
+    ///
+    /// The output stream will not output the same record except during failover and recovery.
+    /// However, if you want to delete one record permanently, you should call
+    /// [`SortBuffer::delete`] manually.
+    pub fn consume(&mut self) -> impl Iterator<Item = DataChunk> + '_ {
```
Comment (st1page) [inline src/stream/src/executor/sort_buffer.rs:153]: should we delete the records in state table here 🤔

**Review by st1page — APPROVED** [review]
generally LGTM.

File: src/stream/src/executor/sort_buffer.rs (hunk 1 above)
Comment (TennyZhuang) [reply src/stream/src/executor/sort_buffer.rs:74]: TBD later

File: src/stream/src/executor/sort_buffer.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/stream/src/executor/sort_buffer.rs:153]: I guess it's hard for caller to use

**Pull request merged**